### PR TITLE
Add cluster features to scattermapbox trace

### PR DESCRIPTION
**Pull request by elben10**
I've added a trace that allow one to use [mapbox clustering layer](https://docs.mapbox.com/mapbox-gl-js/example/cluster/).

Currently it doesn't support the fill, selection and text features.

### CODE REVIEW

**Comment by archmoj** [comment]
@elben10 Thanks for the PR!
Could you please pick this commit https://github.com/plotly/plotly.js/commit/d9c6e02b03f8b9656b452e632c0dc28be9a52c49 that fixes the `mapbox` filename, adds the baseline and some tests.

File: src/traces/scatterclustermapbox/attributes.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            sizemin: markerAttrs.sizemin,
+            sizemode: markerAttrs.sizemode
+        },
+    colorScaleAttrs('marker')
+  ),
+    below: {
+        valType: 'string',
+        role: 'info',
+        description: [
+            'Determines if this scattermapbox trace\'s layers are to be inserted',
```
Comment (archmoj) [inline src/traces/scatterclustermapbox/attributes.js:126]: Shouldn't these here be `scatterclustermapbox`?

Comment (elben10) [reply src/traces/scatterclustermapbox/attributes.js:126]: Yes. It is fixed in the latest commit.

File: src/traces/scatterclustermapbox/convert.js [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        for(var idx = 1; idx < stepSize.length; idx++) {
+            sizeArray.push(steps[idx - 1], stepSize[idx]);
+        }
+        return sizeArray;
+    } else {
+        return stepSize;
+    }
+}
+
+function makeCircleOpts(calcTrace) {
```
Comment (archmoj) [inline src/traces/scatterclustermapbox/convert.js:140]: This function is pretty much identical to the one found in `scattermapbox`.
Let's create a new file at `src/traces/scattermapbox/helpers.js` 
Then export this function 
```js
exports.makeCircleOpts = function (calcTrace) {
...
};
```
And require it here and in `scattermapbox` please.

Comment (elben10) [reply src/traces/scatterclustermapbox/convert.js]: Fixed in latest commit.

File: src/traces/scatterclustermapbox/attributes.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var hovertemplateAttrs = require('../../plots/template_attributes')
+  .hovertemplateAttrs;
+var texttemplateAttrs = require('../../plots/template_attributes')
+  .texttemplateAttrs;
```
Comment (archmoj) [inline src/traces/scatterclustermapbox/attributes.js:14]: Let's avoid these extra lines here and simply
```js
var hovertemplateAttrs = require('../../plots/template_attributes').hovertemplateAttrs;
var texttemplateAttrs = require('../../plots/template_attributes').texttemplateAttrs;
```

**Comment by alexcjohnson** [comment]
@elben10 thanks for the PR, it looks really useful! But let's take a step back before going further with it - does this really need to be a separate trace type? It seems to me as though, in terms of the attributes involved, it's just a few extra options on top of `scattermapbox` to control the clustering, and the attributes regarding lines are not allowed. I'd suggest adding a `cluster.enabled` attribute, which defaults to `true` whenever any of the other `cluster` attributes is set explicitly; then when `cluster.enabled===true` the only `mode` we will allow is `markers` and the `line` attributes will automatically not be used. Then in the following steps we can just look at the `cluster.enabled` attribute to determine which pathway to take.

Does that seem like a reasonable way to include this feature?

**Comment by elben10** [comment]
@alexcjohnson I think that sounds like a better choice. I'll update the code accordingly

File: test/jasmine/assets/mock_lists.js [hunk 4]
```diff
@@ -70,6 +70,7 @@ var glMockList = [
 ];
 
 var mapboxMockList = [
+    ['scatterclustermapbox', require('@mocks/mapbox_scattercluster.json')],
```
Comment (archmoj) [inline test/jasmine/assets/mock_lists.js:73]: Please remove this line now that a new trace is not implemented.

File: src/traces/scattermapbox/plot.js [hunk 5]
```diff
@@ -101,29 +101,179 @@ proto.update = function update(calcTrace) {
...
+            this.setSourceData(k, opts);
+            subplot.setOptions(this.layerIds[k], 'setPaintProperty', opts.paint);
+        }
     }
 
+    // link ref for quick update during selections
+    calcTrace[0].trace._glTrace = this;
+};
+
+scatterclustermapboxproto.dispose = function dispose() {
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:233]: Now that there are two prototypes to take care of, one should check if reacting `scattermapbox` with/without `cluster` works as expected.

Comment (elben10) [reply src/traces/scattermapbox/plot.js:233]: So basically, we need to ensure that the correct layers in mapbox is rendered depending on whether cluster has been assigned a value. Is that correctly understood?

Comment (archmoj) [reply src/traces/scattermapbox/plot.js:233]: Exactly.
Here is a [codepen](https://codepen.io/MojtabaSamimi/pen/WNwKxeQ?editors=0010) illustrating the problem.
We should see the cluster after `Plotly.react` not the initial figure.

File: src/traces/scattermapbox/convert.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
-*
-* This source code is licensed under the MIT license found in the
-* LICENSE file in the root directory of this source tree.
-*/
+ * Copyright 2012-2020, Plotly, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:7]: Please revert this header change.

File: src/traces/scattermapbox/convert.js [hunk 7]
```diff
@@ -61,7 +89,7 @@ module.exports = function convert(gd, calcTrace) {
         fill.layout.visibility = 'visible';
 
         Lib.extendFlat(fill.paint, {
-            'fill-color': trace.fillcolor
+            'fill-color': trace.fillcolor,
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:92]: Would you mind reverting any minor syntax changes & refactors like this one in this PR?
That way it would be easier to keep track of the actual changes required to implement the new feature.
Thank you!

Comment (archmoj) [reply src/traces/scattermapbox/convert.js:92]: Also you are welcome to suggest adding new lint rules here: https://github.com/plotly/plotly.js/issues/3162

File: src/traces/scattermapbox/convert.js [hunk 8]
```diff
@@ -89,7 +89,7 @@ module.exports = function convert(gd, calcTrace) {
         fill.layout.visibility = 'visible';
 
         Lib.extendFlat(fill.paint, {
-            'fill-color': trace.fillcolor,
+            'fill-color': trace.fillcolor
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js]: There is still a number of refactoring changes in this PR that we would like to avoid for this implementation.
Please see the image below showing some (not all) of them.
Screenshot from 2020-09-22 09-44-56

You could view the rest using https://github.com/plotly/plotly.js/pull/5147/files

Comment (elben10) [reply src/traces/scattermapbox/convert.js]: I'll fix those changes tomorrow

File: src/traces/scattermapbox/plot.js (hunk 5 above)
Comment (archmoj) [reply src/traces/scattermapbox/plot.js]: This is now fixed by 3f6ce61.
Thanks @elben10.

File: src/traces/scattermapbox/convert.js (hunk 8 above)
Comment (elben10) [reply src/traces/scattermapbox/convert.js]: It's fixed now

File: src/traces/scattermapbox/plot.js [hunk 9]
```diff
@@ -56,72 +74,108 @@ proto.setSourceData = function(k, opts) {
...
     var trace = calcTrace[0].trace;
-    var scatterMapbox = new ScatterMapbox(subplot, trace.uid);
+    var scatterMapbox = new ScatterMapbox(
+    subplot,
+    trace.uid,
+    trace.cluster.enabled
+  );
     var optsAll = convert(subplot.gd, calcTrace);
-    var below = scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid];
+    var below = (scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid]);
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:162]: Please drop the parentheses here:
```suggestion
    var below = scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid];
```

File: src/traces/scattermapbox/attributes.js [hunk 10]
```diff
@@ -26,6 +26,54 @@ module.exports = overrideAll({
     lon: scatterGeoAttrs.lon,
     lat: scatterGeoAttrs.lat,
 
+    cluster: {
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:29]: We should add a new attribute for enabling cluster option here. Something like:
```js
enabled: {
    valType: 'boolean',
    role: 'info',
    dflt: false,
    editType: 'plot',
    description: 'Determines whether clustering is enabled or disabled.'
},
```

File: src/traces/scattermapbox/attributes.js [hunk 11]
```diff
@@ -26,6 +26,54 @@ module.exports = overrideAll({
...
+        maxZoom: {
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:30]: Please rename this attribute to only use lowercase i.e. `maxzoom`.

File: src/traces/scattermapbox/defaults.js [hunk 12]
```diff
@@ -28,6 +28,13 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         return;
     }
 
+    coerce('cluster.maxZoom');
+    coerce('cluster.radius');
+    coerce('cluster.steps');
+    coerce('cluster.size');
+    coerce('cluster.color');
+    traceOut.cluster.enabled = traceIn.cluster ? true : false;
```
Comment (archmoj) [inline src/traces/scattermapbox/defaults.js:36]: This is not quite right at the moment e.g. it is not going to work if the `cluster` is declared via `template.
But after adjusting the new attributes, it should look like this.
```js
var clusterEnabled = coerce('cluster.enabled');
if(clusterEnabled) {
    coerce('cluster.maxzoom');
    coerce('cluster.radius');
    coerce('cluster.steps');
    coerce('cluster.size');
    coerce('cluster.color');
}
```

File: src/traces/scattermapbox/plot.js [hunk 13]
```diff
@@ -42,11 +50,21 @@ function ScatterMapbox(subplot, uid) {
...
+            cluster: true,
+            clusterMaxZoom: cluster.maxZoom,
+            clusterRadius: cluster.radius,
+        });
+    } else {
+        this.subplot.map.addSource(this.sourceIds[k], {
+            type: 'geojson',
+            data: opts.geojson,
+        });
+    }
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:67]: The changes here are looking good. Thanks for merging two prototypes.
But we could still improve it a bit.
```js
proto.addSource = function(k, opts, cluster) {
    var sourceOpts =
            type: 'geojson',
            data: opts.geojson,
    };
    
    if(cluster.enabled) {
        Lib.extendFlat({}, sourceOpts, {
            cluster: true, // better to rename this key to clusterEnabled?
            clusterMaxZoom: cluster.maxZoom,
            clusterRadius: cluster.radius,
        });
    } 
    
    this.subplot.map.addSource(this.sourceIds[k], sourceOpts);
};
```

File: src/traces/scattermapbox/defaults.js [hunk 14]
```diff
@@ -28,6 +28,15 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         return;
     }
 
+    var clusterEnabled = coerce('cluster.enabled');
```
Comment (archmoj) [inline src/traces/scattermapbox/defaults.js:31]: You should add `enabled` to the API as well. Please see my comment here: https://github.com/plotly/plotly.js/pull/5147/files#r494316347

File: src/traces/scattermapbox/hover.js [hunk 15]
```diff
@@ -34,6 +42,7 @@ module.exports = function hoverPoints(pointData, xval, yval) {
     function distFn(d) {
         var lonlat = d.lonlat;
         if(lonlat[0] === BADNUM) return Infinity;
+        if(trace.cluster.enabled && !clusteredPointsIds.includes(d.i + 1)) return Infinity;
```
Comment (archmoj) [inline src/traces/scattermapbox/hover.js:45]: `includes` is not suported in IE.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes
You may use `indexOf` instead:
```js
if(trace.cluster.enabled && clusteredPointsIds.indexOf(d.i + 1) === -1) return Infinity;
```

File: src/traces/scattermapbox/convert.js [hunk 16]
```diff
@@ -331,3 +357,32 @@ function blankFillFunc() { return ''; }
 function isBADNUM(lonlat) {
     return lonlat[0] === BADNUM;
 }
+
+function createClusterColors(colors, steps) {
+    var idx, colors_;
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:362]: Non-blocking:
But it would be great if you rename `colors_` temporary variable to `_colors` or even better `newColors`.
Similarly `sizes_` > `newSizes`

File: src/traces/scattermapbox/plot.js [hunk 17]
```diff
@@ -10,25 +10,33 @@
 
 var convert = require('./convert');
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:11]: Please add
```js
var Lib = require('../../lib');
```
here.

File: src/traces/scattermapbox/plot.js [hunk 18]
```diff
@@ -42,11 +50,21 @@ function ScatterMapbox(subplot, uid) {
...
-        data: opts.geojson
-    });
+proto.addSource = function(k, opts, cluster) {
+    var sourceOpts = {
+            type: 'geojson',
+            data: opts.geojson,
+    };
+    
+    if(cluster.enabled) {
+        Lib.extendFlat({}, sourceOpts, {
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:60]: My apologies. Please remove the first argument to this function.
This line should be
```js
Lib.extendFlat(sourceOpts, {
```

File: test/image/mocks/mapbox_scattercluster.json [hunk 19]
```diff
@@ -0,0 +1,37 @@
+{
+    "data": [
+      {
+        "type": "scattermapbox",
+        "mode": "markers",
+        "cluster": {},
```
Comment (archmoj) [inline test/image/mocks/mapbox_scattercluster.json:6]: Here we should enable clustering.
```json
        "cluster": { "enabled": true },
```

File: src/traces/scattermapbox/convert.js [hunk 20]
```diff
@@ -32,19 +32,39 @@ module.exports = function convert(gd, calcTrace) {
...
+                'circle-radius': createClusterSizes(trace.cluster.size, trace.cluster.steps)
+            },
+        },
+        clusterCount: {
+            type: 'symbol',
+            filter: ['has', 'point_count'],
+            paint: {},
+            layout: {
+                'text-field': '{point_count_abbreviated}',
+                'text-font': ['DIN Offc Pro Medium', 'Arial Unicode MS Bold'],
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:64]:
```suggestion
                'text-font': ['Open Sans Regular', 'Arial Unicode MS Regular'], // reuse dflt found in mapbox/layout_attributes.js 
```
It might be better to use similar defaults found here: 
https://github.com/plotly/plotly.js/blob/86f5e2d02d56a94fcf2c26b30d3539fd54211c40/src/plots/mapbox/layout_attributes.js#L27

File: src/traces/scattermapbox/attributes.js [hunk 21]
```diff
@@ -26,6 +26,61 @@ module.exports = overrideAll({
...
+            role: 'info',
+            valType: 'number',
+            arrayOk: true,
+            dflt: -1,
+            min: -1,
+            description: [
+                'Sets the steps for each cluster.'
+            ].join(' ')
+        },
+        color: {
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:61]: It seems we could drop `color` and `size` here and reuse `trace.marker.color` and `trace.marker.size`.

File: src/traces/scattermapbox/hover.js [hunk 22]
```diff
@@ -34,6 +42,7 @@ module.exports = function hoverPoints(pointData, xval, yval) {
     function distFn(d) {
         var lonlat = d.lonlat;
         if(lonlat[0] === BADNUM) return Infinity;
+        if(trace.cluster.enabled && clusteredPointsIds.indexOf(d.i + 1) === -1) return Infinity;
```
Comment (archmoj) [inline src/traces/scattermapbox/hover.js:45]: The hover does not appear!
But if you revert the changes if `hover.js` it displays the original data points; which I think is desirable.

File: src/traces/scattermapbox/plot.js [hunk 23]
```diff
@@ -56,72 +75,108 @@ proto.setSourceData = function(k, opts) {
...
     var optsAll = convert(subplot.gd, calcTrace);
     var below = subplot.belowLookup['trace-' + this.uid];
-    var i, k, opts;
-
-    if(below !== this.below) {
-        for(i = ORDER.length - 1; i >= 0; i--) {
-            k = ORDER[i];
+    var i, k, opts, order;
+
+    if(trace.cluster.enabled === this.clusterEnabled) {
```
Comment (archmoj) [inline src/traces/scattermapbox/plot.js:99]: To make this part more readable, I suggest declaring two variables like this and use them everywhere below.

```js
var hasCluster = trace.cluster.enabled;
var hadCluster = this.clusterEnabled;
if(hasCluster === hadCluster) {
```

File: src/traces/scattermapbox/convert.js [hunk 24]
```diff
@@ -90,6 +110,10 @@ module.exports = function convert(gd, calcTrace) {
         });
     }
 
+    if(hasCircles && hasCluster) {
+        circle.filter = ['!', ['has', 'point_count']];
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:114]: Non-blocking:
You could put this inside the above  is statement:

```js
if(hasCircles) {
        var circleOpts = makeCircleOpts(calcTrace);
        circle.geojson = circleOpts.geojson;
        circle.layout.visibility = 'visible';
        if(hasCluster) circle.filter = ['!', ['has', 'point_count']];

        Lib.extendFlat(circle.paint, {
            'circle-color': circleOpts.mcc,
            'circle-radius': circleOpts.mrc,
            'circle-opacity': circleOpts.mo
        });
    }
```

File: src/traces/scattermapbox/convert.js [hunk 25]
```diff
@@ -32,19 +32,39 @@ module.exports = function convert(gd, calcTrace) {
...
+    var circle = initContainer('circle');
+    var symbol = initContainer('symbol');
 
     var opts = {
         fill: fill,
         line: line,
         circle: circle,
-        symbol: symbol
+        symbol: symbol,
+        cluster: {
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:49]: We should possibly move these `cluster` and `clusterCount` below and put them inside `if(hasCircles) {` block.

File: src/traces/scattermapbox/hover.js (hunk 22 above)
Comment (elben10) [reply src/traces/scattermapbox/hover.js:45]: The hover functionality should at the moment work when the individual points is unpacked from the cluster. 

I think it is desirable to enable the hover functionality is being enabled when the individual points is unpacked from its cluster

File: src/traces/scattermapbox/attributes.js (hunk 21 above)
Comment (elben10) [reply src/traces/scattermapbox/attributes.js]: I don't think we can drop color and size. Shouldn't one be able to have different color and sized point and clusters. If we use the color from `trace.marker.color` it will no be possible to have a cluster with say 3 different colors, and the points being pained with another color

**Review by archmoj — COMMENTED** [review]
Thanks @Jakob Jul Elben.
We are getting close.
Here are few comments to consider.
I'll also ask @nicolaskruchten and @alexcjohnson to review the API and implementation.

File: src/traces/scattermapbox/defaults.js (hunk 14 above)
Comment (archmoj) [inline src/traces/scattermapbox/defaults.js:31]: @alexcjohnson should we use `Lib.coerce2` for this attribute?

File: src/traces/scattermapbox/attributes.js [hunk 26]
```diff
@@ -26,6 +26,61 @@ module.exports = overrideAll({
...
+            description: 'Radius of each cluster when clustering points.',
+        },
+        steps: {
+            role: 'info',
+            valType: 'number',
+            arrayOk: true,
+            dflt: -1,
+            min: -1,
+            description: [
+                'Sets the steps for each cluster.'
```
Comment (nicolaskruchten) [inline src/traces/scattermapbox/attributes.js:58]: I'm confused about what this attribute does?

File: src/traces/scattermapbox/attributes.js (hunk 21 above)
Comment (nicolaskruchten) [reply src/traces/scattermapbox/attributes.js]: I think instead of having color and size both be arrays, we might want to have a single attribute that accepts an array of color/size tuples like `steps: [ {color: "red", size: 10} ]` or something?

File: src/traces/scattermapbox/attributes.js (hunk 26 above)
Comment (elben10) [reply src/traces/scattermapbox/attributes.js:58]: The cluster can be coloured in different colors and take different sizes dependent on how many observations that are contained in a cluster. The steps determines when the cluster will change size or/and color

File: src/traces/scattermapbox/attributes.js (hunk 21 above)
Comment (elben10) [reply src/traces/scattermapbox/attributes.js]: So we need to have either a `color`/`size` to be a singular value or an array. If `color`/`size` is an array, steps should also be defined as an array that has length one less than `color`/`size`. `steps` defines at which levels the cluster changes `color`/`size` dependent on how many observations the cluster contains. 

One way to achieve that would be

`steps: [ {color: "red", size: 10, "step": null}, {color: "blue", size: 20, "step": 20} ]`

In this case, the cluster will be coloured red and sized 10 as long as there are less than 20 observations in the cluster else the cluster will be coloured blue and sized 20.

**Comment by archmoj** [comment]
@elben10 for your info I added this pull to the `v1.58.0` milestone.

File: test/image/mocks/mapbox_scattercluster.json [hunk 27]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (archmoj) [inline test/image/mocks/mapbox_scattercluster.json:1]: Let's improve this mock using the figure presented in this [demo](https://codepen.io/MojtabaSamimi/pen/xxOVLxK?editors=1000).
Thanks.

File: src/traces/scattermapbox/attributes.js (hunk 26 above)
Comment (archmoj) [reply src/traces/scattermapbox/attributes.js:58]: I didn't notice any changes when trying to tweak this parameter.
@elben10 wondering if you could provide an illustration?

File: src/traces/scattermapbox/attributes.js [hunk 28]
```diff
@@ -26,6 +26,58 @@ module.exports = overrideAll({
...
+            dflt: 20,
+            min: 0,
+            description: [
+                'Sets the size for each cluster step.'
+            ].join(' ')
+        },
+        color: {
+            role: 'info',
+            valType: 'color',
+            arrayOk: true,
```
[the remark this answers is not in the document]
Comment (archmoj) [reply src/traces/scattermapbox/attributes.js:74]: I tested this. In fact `cluster.color` cannot be an `arrayOk` attribute.
So please drop this line:
```js
arrayOk: true,
```

File: src/traces/scattermapbox/attributes.js [hunk 29]
```diff
@@ -26,6 +26,58 @@ module.exports = overrideAll({
...
+        size: {
+            role: 'info',
+            valType: 'number',
+            arrayOk: true,
+            dflt: 20,
+            min: 0,
+            description: [
+                'Sets the size for each cluster step.'
+            ].join(' ')
+        },
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:70]: In addition to `cluster.color`, it would be nice to add `cluster.opacity` too. That way one could easily adjust transparency for those large clustered regions over the map.

File: src/traces/scattermapbox/defaults.js [hunk 30]
```diff
@@ -28,6 +28,15 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
...
+    if(clusterEnabled) {
+        coerce('cluster.maxzoom');
+        coerce('cluster.radius');
+        coerce('cluster.step');
+        coerce('cluster.color', defaultColor);
+        coerce('cluster.size');
```
Comment (archmoj) [inline src/traces/scattermapbox/defaults.js:37]: Let's also consider using `marker.color` and `marker.size` as the `default` second argument passed to `coerce`.
For that we need to move this block below `coerce(marker.`*.

Something like:
```js
var marker;
if(subTypes.hasMarkers(traceOut)) {
    handleMarkerDefaults(traceIn, traceOut, defaultColor, layout, coerce, {noLine: true});
    coerce('marker.allowoverlap');
    coerce('marker.angle');
    // array marker.size and marker.color are only supported with circles
    marker = traceOut.marker;
    if(marker.symbol !== 'circle') {
        if(Lib.isArrayOrTypedArray(marker.size)) marker.size = marker.size[0];
        if(Lib.isArrayOrTypedArray(marker.color)) marker.color = marker.color[0];
    }
}

var clusterEnabled = coerce('cluster.enabled');
if(clusterEnabled) {
    coerce('cluster.color', marker.color || defaultColor);
    var clusterSize = coerce('cluster.size', marker.size);
    coerce('cluster.radius', clusterSize);
    coerce('cluster.maxzoom');
    coerce('cluster.step'); // if necessary
}
```

File: test/jasmine/assets/scatter_cluster_mapbox_test.js [hunk 31]
```diff
@@ -0,0 +1,176 @@
+var Plots = require('@src/plots/plots');
+var Lib = require('@src/lib');
+var Axes = require('@src/plots/cartesian/axes');
+
+var ScatterClusterMapbox = require('@src/traces/scatterclustermapbox');
```
Comment (archmoj) [inline test/jasmine/assets/scatter_cluster_mapbox_test.js:5]: A trace titled `scatterclustermapbox` no longer exists.
Please integrate these tests with `scattermapbox` tests.
Thank you!

File: src/traces/scattermapbox/attributes.js [hunk 32]
```diff
@@ -26,6 +26,58 @@ module.exports = overrideAll({
...
 
+    cluster: {
+        enabled: {
+            valType: 'boolean',
+            role: 'info',
+            dflt: false,
+            editType: 'plot',
+            description: 'Determines whether clustering is enabled or disabled.'
+        },
+        maxzoom: {
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:37]: We could possibly drop this attribute and instead reuse `layout.mapbox.(min|max)zoom`, right?
https://github.com/plotly/plotly.js/blob/39fe198ad844e62432d30e11b862b34ef9b1f193/src/plots/mapbox/layout_attributes.js#L241-L251
If not, we should probably make `max`, `dflt` and `description` values similar to those of `mapbox.maxzoom`.

File: src/traces/scattermapbox/attributes.js [hunk 33]
```diff
@@ -26,6 +26,58 @@ module.exports = overrideAll({
     lon: scatterGeoAttrs.lon,
     lat: scatterGeoAttrs.lat,
 
+    cluster: {
+        enabled: {
+            valType: 'boolean',
+            role: 'info',
+            dflt: false,
+            editType: 'plot',
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:34]: Please remove this line:
```js
editType: 'plot',
```
For all these attributes, the `editType` is going to be set to `calc` down below.

File: src/traces/scattermapbox/attributes.js (hunk 32 above)
Comment (alexcjohnson) [reply src/traces/scattermapbox/attributes.js:37]: That's `layout.mapbox.layers[i].(min|max)zoom` and determines the visibility of just that layer. This has a similar effect, but only for this one trace, so seems necessary to me.

Curious though why they have different `max` values (22 vs 24)?

**Comment by elben10** [comment]
> @elben10 for your info I added this pull to the `v1.58.0` milestone.

Nice. I'll fix the last changes during the week

File: src/traces/scattermapbox/attributes.js [hunk 34]
```diff
@@ -26,6 +26,58 @@ module.exports = overrideAll({
...
+            min: -1,
+            description: [
+                'Sets the steps for each cluster.'
+            ].join(' ')
+        },
+        size: {
+            role: 'info',
+            valType: 'number',
+            arrayOk: true,
+            dflt: 20,
```
[the remark this answers is not in the document]
Comment (elben10) [reply src/traces/scattermapbox/attributes.js:64]: Here is an example illustrating why the cluster need a size attribute different from the markers. The marker size sets the size of the markers when the cluster dissolves. The cluster size attributes determines the size of the clusters.

In the example posted below, it means that if there is more than 3 points in the cluster the cluster will have a size of 40. If the there is fewer than 3, it will have the size 20.

https://codepen.io/elben10/pen/yLJRpYQ

File: src/traces/scattermapbox/attributes.js [hunk 35]
```diff
@@ -26,6 +27,51 @@ module.exports = overrideAll({
...
     lat: scatterGeoAttrs.lat,
 
+    cluster: {
+        enabled: {
+            valType: 'boolean',
+            role: 'info',
+            dflt: false,
+            description: 'Determines whether clustering is enabled or disabled.'
+        },
+        maxzoom: mapboxLayoutAtributes.maxzoom,
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:37]: `maxzoom` is inside `layers` and we also want to override the description.
So, this line should become:
```js
maxzoom: extendFlat({}, mapboxLayoutAtributes.layers.maxzoom, {
    description: [
        'Sets the maximum zoom level.',
        'At zoom levels equal to or greater than the maxzoom, the layer will be hidden.'
    ].join(' ')
}),
```

File: src/traces/scattermapbox/attributes.js [hunk 36]
```diff
@@ -18,6 +18,7 @@ var colorScaleAttrs = require('../../components/colorscale/attributes');
 
 var extendFlat = require('../../lib/extend').extendFlat;
 var overrideAll = require('../../plot_api/edit_types').overrideAll;
+var mapboxLayoutAtributes = require('../../plots/mapbox/layout_attributes')
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:21]: Missing a semicolon at the end of this line.
```
npm run test-syntax
```
could also catch these.

File: src/traces/scattermapbox/attributes.js [hunk 37]
```diff
@@ -26,6 +27,51 @@ module.exports = overrideAll({
...
+        },
+        color: {
+            valType: 'color',
+            arrayOk: true,
+            role: 'style',
+            description: [
+                'Sets the color for each cluster step.'
+            ].join(' ')
+        },
+        opacity: markerAttrs.opacity,
```
Comment (archmoj) [inline src/traces/scattermapbox/attributes.js:72]: We should provide a `dflt` for this attribute:
```js
opacity: extendFlat({}, markerAttrs.opacity, {
   dflt: 1
})
```
otherwise the cluster points would be invisible.

File: src/traces/scattermapbox/convert.js [hunk 38]
```diff
@@ -82,6 +83,29 @@ module.exports = function convert(gd, calcTrace) {
...
+        if(hasCluster) {
+            circle.filter = ['!', ['has', 'point_count']];
+            opts.cluster = {
+                type: 'circle',
+                filter: ['has', 'point_count'],
+                layout: {visibility: 'visible'},
+                paint: {
+                    'circle-color': createClusterColors(trace.cluster.color, trace.cluster.step),
+                    'circle-radius': createClusterSizes(trace.cluster.size, trace.cluster.step),
+                    'circle-opacity': trace.cluster.opacity
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:95]: Since the `opacity` is `arrayOk`, this line should be something like:
```js
'circle-opacity': createClusterOpacities(trace.cluster.opacity, trace.cluster.step),
```

File: src/traces/scattermapbox/attributes.js (hunk 36 above)
Comment (elben10) [reply src/traces/scattermapbox/attributes.js:21]: Thanks

File: src/traces/scattermapbox/attributes.js (hunk 26 above)
Comment (archmoj) [reply src/traces/scattermapbox/attributes.js:56]: > The cluster can be coloured in different colors and take different sizes dependent on how many observations that are contained in a cluster. The steps determines when the cluster will change size or/and color

Thanks @elben10 please provide part of this info in the attribute description.

File: src/traces/scattermapbox/convert.js [hunk 39]
```diff
@@ -331,3 +361,48 @@ function blankFillFunc() { return ''; }
 function isBADNUM(lonlat) {
     return lonlat[0] === BADNUM;
 }
+
+function createClusterColors(colors, step) {
+    var idx, newColors;
+    if(Lib.isArrayOrTypedArray(colors)) {
```
Comment (archmoj) [inline src/traces/scattermapbox/convert.js:367]: Here we should also check for `step` to be an array.
On another note. Wondering if you could merge/reuse all/any of these 3 functions: `createClusterColors`, `createClusterSizes` and `createClusterOpacities`?

File: src/traces/scattermapbox/defaults.js (hunk 30 above)
Comment (archmoj) [reply src/traces/scattermapbox/defaults.js:37]: Above comment is not addressed at the moment.

File: test/jasmine/assets/scatter_cluster_mapbox_test.js (hunk 31 above)
Comment (archmoj) [reply test/jasmine/assets/scatter_cluster_mapbox_test.js]: @elben10  could explain why the interactive tests are removed in https://github.com/plotly/plotly.js/pull/5147/commits/1f42ee217b321134740505010e374f707a27c1d1?

Comment (elben10) [reply test/jasmine/assets/scatter_cluster_mapbox_test.js]: The tests should be integrated in [`scattermapbox_test.js`](https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/scattermapbox_test.js) now where scattermapboxcluster doesn't exist as a separate trace anymore. Or am I missing something? (I am aware, that these test is not integrated yet)

Comment (archmoj) [reply test/jasmine/assets/scatter_cluster_mapbox_test.js]: > The tests should be integrated in [`scattermapbox_test.js`](https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/scattermapbox_test.js) now where scattermapboxcluster doesn't exist as a separate trace anymore. Or am I missing something? (I am aware, that these test is not integrated yet)

Thanks very much for the note and hard work on this feature. 
This PR seems almost complete with jasmine tests.

File: src/traces/scattermapbox/convert.js (hunk 39 above)
Comment (elben10) [reply src/traces/scattermapbox/convert.js:367]: I'm pretty sure it is possible. I'll take a look at it

**Comment by archmoj** [comment]
To avoid merge conflicts, before adding jasmine tests, could you fetch and merge upstream master into this branch please?

**Comment by archmoj** [comment]
337 file changes! This is now extremely hard to review.
Could you start a fresh PR and rebase/squash the  changes you made to mapbox only on top of v1.58.1?
Please also consider giving us as the maintainers permission to edit as described in the PR template. 
Thank you!

**Comment by archmoj** [comment]
@elben10 
To make it easier for you to submit a new PR I fixed the merge conflicts on this temporary branch:
https://github.com/plotly/plotly.js/compare/cluster-scattermapbox

To open the final PR you may consider starting a new fork.
Then
```
git checkout cluster-scattermapbox
git reset HEAD^
git add .
git commit 
``` 
So that you get credit for the great contribution you made.
Then 
```
git push -f
```
Then please open a PR from your `cluster-scattermapbox` branch to our `master`.

Does that make sense?

**Comment by elben10** [comment]
Yes it does. The pull request is now created! Thanks for the help and guidance!
